exp/addons/controllers/clusterresourceset_controller.go

```diff
@@ -0,0 +1,407 @@
+			for key := range resourceSecret.Data {
+				keys = append(keys, key)
+			}
+			sort.Strings(keys)
+
+			for _, key := range keys {
+				dataList = append(dataList, resourceSecret.Data[key])
+			}
+		default:
+			logger.Error(err, "Unsupported ClusterResourceSet resource kind", "Resource Kind", resource.Kind)
+			conditions.MarkFalse(clusterResourceSet, addonsv1.ResourcesAppliedCondition, addonsv1.UnsupportedResourceTypeReason, clusterv1.ConditionSeverityWarning, "")
+			errList = append(errList, err)
+			continue
+		}
+
+		// apply all values in the key-value pair of the resource to the cluster.
+		// As there can be multiple key-value pairs in a resource, each value may have multiple objects in it.
+		isSuccessful := true
+		for i := range dataList {
+			data := dataList[i]
+
+			if err := apply(ctx, remoteClient, data); err != nil {
+				isSuccessful = false
+				logger.Error(err, "failed to apply ClusterResourceSet resource", "Resource kind", resource.Kind, "Resource name", resource.Name)
+				conditions.MarkFalse(clusterResourceSet, addonsv1.ResourcesAppliedCondition, addonsv1.ApplyFailedReason, clusterv1.ConditionSeverityWarning, err.Error())
+				errList = append(errList, err)
+			}
+		}
+
+		binding = generateResourceStatus(isSuccessful, computeHash(dataList))
+		bindings.Resources[resourceKindName] = binding
+	}
+	if len(errList) > 0 {
+		return kerrors.NewAggregate(errList)
+	}
+
+	conditions.MarkTrue(clusterResourceSet, addonsv1.ResourcesAppliedCondition)
+
+	return nil
+}
+
+// patchOwnerRefToResource patches the  ClusterResourceSet as a OwnerReference to the resource.
+func (r *ClusterResourceSetReconciler) patchOwnerRefToResource(ctx context.Context, clusterResourceSet *addonsv1.ClusterResourceSet, resource *unstructured.Unstructured) error {
+	patch := client.MergeFrom(resource.DeepCopy())
+
+	newRef := metav1.OwnerReference{
+		APIVersion: clusterResourceSet.GroupVersionKind().GroupVersion().String(),
+		Kind:       clusterResourceSet.GroupVersionKind().Kind,
+		Name:       clusterResourceSet.GetName(),
+		UID:        clusterResourceSet.GetUID(),
+	}
+
+	refs := resource.GetOwnerReferences()
+	if !util.IsOwnedByObject(resource, clusterResourceSet) {
+		refs = append(refs, newRef)
+		resource.SetOwnerReferences(refs)
+		return r.Client.Patch(ctx, resource, patch)
+	}
+	return nil
+}
+
+// generateResourceStatus returns ResourceBinding to  be used in ClusterResourceSetBinding resource of the matching cluster
+func generateResourceStatus(isSuccessful bool, hash string) addonsv1.ResourceBinding {
+	return addonsv1.ResourceBinding{
+		Hash:            hash,
+		Applied:         isSuccessful,
+		LastAppliedTime: &metav1.Time{Time: time.Now().UTC()},
+	}
+}
+
+// generateResourceKindName generates a unique name to identify resources that is used in resources map in ClusterResourceSetBinding.
+func generateResourceKindName(resource addonsv1.ResourceRef) string {
+	return resource.Kind + "/" + resource.Name
+}
```

Comment:
If we move to the slice approach, we can avoid creating new naming conventions

Reply:
Opened another issue: #3297